MiniApp Widget specification draft

lumanz opened the pull request.

lumanz committed:
new spec

chenyinli committed:
update spec

lumanz committed:
Merge pull request #1 from chenyinli/main

update spec to use respec

xfq committed:
Trigger Netlify

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+
+  <link rel="stylesheet" href="local.css" />
+
+  <style>
+    table {
+      border-collapse: collapse;
+    }
+
+    table,
+    th,
+    td {
+      border: 1px solid #666;
+    }
+
+    td {
+      padding: 2px 15px;
+    }
+  </style>
```
This can be removed.

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+        company: "Xiaomi",
+        w3cid: 124040
+      }, {
+        name: "Xiangyang Yao",
+        companyURL: "https://www.mi.com/",
+        company: "Xiaomi"
+      }, {
+        name: "Wen Pan",
+        companyURL: "https://www.mi.com/",
+        company: "Xiaomi"
+      }, {
+        name: "Hai Jiang",
+        companyURL: "https://www.mi.com/",
+        company: "Xiaomi"
```
Are these three editors members of the MiniApps CG? We need to ensure that every editor/contributor is a member of the community group for IPR reasons.

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+      与页面不同，widget可以仅占用屏幕的部分区域（而不是全部），用于为用户显示关键信息和响应简单的用户操作，例如显示用户所在地的天气，跟踪用户当前的行程状态并提供进一步的操作。
+    </p>
+    <p its-locale-filter-list="en" lang="en">
+      This specification defines addressing,packaging, manifest and lifecycle for a MiniApp Widget. The interaction and communication between MiniApp widget and MiniApp or host environment are also defined.
```
```suggestion
      This specification defines addressing, packaging, manifest and lifecycle for a MiniApp Widget. The interaction and communication between MiniApp widget and MiniApp or host environment are also defined.
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+
+  <script class="remove">
+    var respecConfig = {
+      specStatus: "ED",
```
```suggestion
      specStatus: "CG-DRAFT",
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+      edDraftURI: "https://w3c.github.io/miniapp-widget/sepc/",
+
+      shortName: "miniapp-widget-spec",
+      noRecTrack: true,
```
This line can be removed.

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+        company: "Xiaomi"
+      }],
+
+      group: "wg/miniapps",
```
```suggestion
      group: "cg/miniapps",
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp widget 是 MiniApp页面的一种特殊形式。与页面不同，widget可以仅占用屏幕的部分区域（而不是全部），
+        用于为用户显示关键信息和响应简单的用户操作，例如显示用户所在地的天气，跟踪用户当前的行程状态并提供进一步的操作。</p>
+  
+      <p>MiniApp widget is a special form of MiniApp page. Unlike pages, widgets can use only part of the screen (rather than all of
```
Note that if we're going to maintain bilingual versions in the future, we'll need to add the `its-locale-filter-list` and `lang` attributes.

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+    </section>
+
+    <section id='terminology'>
+      <h3>Terminology</h3>
```
Some terms are already [defined](https://w3c.github.io/miniapp-packaging/#sec-terminology) in MiniApp Packaging. We should probably not define and export (using `data-export`) them again here.

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+
+  <section id='usecase'>
+    <h2>
+      <span its-locale-filter-list="en" lang="en">Use case</span>
```
```suggestion
      <span its-locale-filter-list="en" lang="en">Use cases</span>
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">用户在浏览视频或者图片类 MiniApp 时，自动识别内容中的商品。当用户点击这个商品时，打开电商 MiniApp 相关的 MiniApp Widget。
+      这个 Widget 可以浮窗形式出现，呈现商品的一些基本信息。用户可以点击Widget 上的详情按钮，跳转到电商 MiniApp 的页面来购买。</p>
+    <p its-locale-filter-list="en" lang="en">In video or picture MiniApps, the products
+      in the content can be identified automatically. When user clicks on this item,
```
```suggestion
      in the content can be identified automatically. When a user clicks on this item,
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+      by <a href="https://w3c.github.io/miniapp-addressing/">Miniapp Addressing</a>.</p>
+
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">与 page 不同的是， MiniApp widget 并不存在 start widget。因此，指定某一个 MiniApp widget 需要显式说明 Widget 的 path，如下例所示：</p>
+    <p its-locale-filter-list="en" lang="en">Unlike Miniapp page, there is no start widget for a MiniApp widget. Therefore, specifying a MiniApp widget needs to
```
```suggestion
    <p its-locale-filter-list="en" lang="en">Unlike MiniApp pages, there is no start widget for a MiniApp widget. Therefore, specifying a MiniApp widget needs to
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+    </h2>
+
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">包含 Widget 的 MiniApp 包结构符合如下要求：</p>
+    <p its-locale-filter-list="en" lang="en">The structure of MiniApp package containing a Widget should meets the following requirements:</p>
```
I suggest using a list for the requirements below, which is a bit clearer than `p`. See [clreq](https://github.com/w3c/clreq/blob/2b8f9ccc80c7731bb1662e05d1aea086a9c82b4b/index.html#L222-L268) for example markup.

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+    <section id="window-size">
+      <h3>Widget obtains windows size</h3>
+    
+      <p its-locale-filter-list="zh-hans" lang="zh-hans"> widget 可以获取窗口大小，方便应用做对应布局。在 MiniApp 中，通过 window.screen 可以获取
```
```suggestion
      <p its-locale-filter-list="zh-hans" lang="zh-hans"> widget 可以获取窗口大小，方便应用做对应布局。在 MiniApp 中，通过 `window.screen` 可以获取
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+        screen信息。</p>
+      <p its-locale-filter-list="en" lang="en"> The widget can get window size. This is convenient for widget to layout
+        correspondingly. In MiniApp, screen information
+        can be obtained through window.screen.</p>
```
```suggestion
        can be obtained through `window.screen`.</p>
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,546 @@
+    </section>
+  </section>
+
+  <section>
+    <h2>Security considerations</h2>
+  </section>
+
+  <section>
+    <h2>Privacy considerations</h2>
+  </section>
```
Just a note that we need to add content in the future.

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update index.html

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update index.html

xfq committed:
Update index.html

xfq committed:
Update spec/index.html

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update index.html

lumanz committed:
Update index.html

lumanz committed:
Update index.html

lumanz committed:
Update index.html

lumanz committed:
Update index.html

lumanz committed:
Update index.html

QingAn commented on spec/index.html:
```diff
@@ -0,0 +1,553 @@
+        when a product widget is displayed on search resualts page, user clicks on
+        another product, the content of the widget will be updated with the infomation of
+        the new product.</p>
+
```
It seems not clear what the strong interaction means here. In the above example "when a product widget is displayed on search results page, user clicks on another product, the content of the widget will be updated with the infomation of the new product", it is more like clicking a product link and then opening the new link, which seems quite usual in Web world. What is the motivation to reuse the existing wiget environment, instead of creating a new one?

Also, suggest to include how to update the widget. Request and fetech new xml, css, js and image resources from server, and replace the existing resources?

And, if user wants to go back to the results page, how to switch back?

QingAn commented on spec/index.html:
```diff
@@ -0,0 +1,553 @@
+<head>
+  <meta charset="utf-8" />
+  <title>MiniApp Widget Specification</title>
+
```
The title seems not consistent with the contents. If we go with "Specification", I think use case section should be removed?

Also, if it is specification, I think for at least lifecycle part, WebIDL format should be defined.

lumanz committed:
Update index.html

lumanz committed:
Use case deleted

lumanz committed:
Update index.html

added update process description in 5.4
added interface and callback in 5.5
added local data and file access in 7.2

lumanz committed:
Update index.html

xfq commented on spec/index.html:
```diff
@@ -362,23 +362,65 @@ <h3>
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp的update事件并不会触发页面状态的变化，页面仍然会保留在‘shown’状态。该事件发生时，会触发onupdate回调。</p>
+      <p its-locale-filter-list="en" lang="en">The update event of MiniApp does not trigger the change of the page state, the page will remain in the 'shown' state. When this event occurs, the onupdate callback is triggered.</p>
+ 
+  <pre class="JavaScript">
```
Should be `<pre class="example js">`.

xfq commented on spec/index.html:
```diff
@@ -362,23 +362,65 @@ <h3>
+  <p its-locale-filter-list="zh-hans" lang="zh-hans">示例如下：</p>
+  <p its-locale-filter-list="en" lang="en">An example is as follows:</p>
+  
+  <pre class="JavaScript">
```
Should be `<pre class="idl">`.

xfq commented on spec/index.html:
```diff
@@ -462,11 +504,89 @@ <h3>
+  <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp Widget和MiniApp页面都可以通过user agent读写在本地保存的数据，没有过期时间。MiniApp Widget和MiniApp可以通过localStorage接口来访问这些数据。示例如下：</p>
+      <p its-locale-filter-list="en" lang="en">Both the MiniApp Widget and the MiniApp page can read and write locally saved data through the user agent with no expiration time. MiniApp Widget and MiniApp can access these data through the localStorage interface. An example is as follows:</p>
+    
+        <pre class="JavaScript">
```
Should be `<pre class="example js">`.

xfq commented on spec/index.html:
```diff
@@ -462,11 +504,89 @@ <h3>
+     <p its-locale-filter-list="en" lang="en">MiniApp Widget and MiniApp page can access files through the src attribute of the tag and the File API:</p>
+     <p its-locale-filter-list="en" lang="en">Example of tag src attribute:</p>
+      
+      <pre class="JavaScript">
```
Should be `<pre class="example js">`.

xfq commented on spec/index.html:
```diff
@@ -462,11 +504,89 @@ <h3>
+     <p its-locale-filter-list="en" lang="en">Example of tag src attribute:</p>
+      
+      <pre class="JavaScript">
+         <image src="platform://miniapp/foo/common/example.png"></image>
```
`<` should be escaped (i.e., `&lt;`).

xfq commented on spec/index.html:
```diff
@@ -462,11 +504,89 @@ <h3>
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">通过File API创建和访问本地文件：</p>
+     <p its-locale-filter-list="en" lang="en">Create and access local files via the File API:</p>   
+
+          <pre class="JavaScript">
```
Should be `<pre class="example js">`.

chenyinli committed:
fix some issues in spec doc

lumanz committed:
Merge pull request #2 from chenyinli/main

fix some issues in spec doc

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,634 @@
+
+    <section id='terminology'>
+      <h3>
+      <span its-locale-filter-list="en" lang="en">Terminology</span>
```
I would suggest that we link to https://www.w3.org/TR/miniapp-packaging/#sec-terminology directly instead of defining the same terms here. Otherwise, we are maintaining two glossaries and synchronizing the definitions on both sides.

espinr replied:
+1. I will update the terms in Packaging if needed.

lumanz committed:
Update index.html

lumanz committed:
Update index.html

xfq commented on spec/index.html:
```diff
@@ -493,26 +493,16 @@ <h3>
-      <p its-locale-filter-list="zh-hans" lang="zh-hans">Widget 和 MiniApp 应该都可以访问 MiniApp 包内的文件资源以及系统为 MiniApp 分配的存储其相关数据和信息的文件。对于同一个id的Widget和MiniApp，访问同一数据和文件的权限保持一致。</p>
-      <p its-locale-filter-list="en" lang="en">Widget and the MiniApp should be able to access the file resources in the MiniApp package and the files
-        allocated by the system for the MiniApp to store its related data and information. For Widgets and MiniApp with the same id, the permissions to access the same data and files remain the same.</p>
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">iniApp Widget和MiniApp页面都可以通过user agent读写在本地保存的数据，没有过期时间。MiniApp Widget和MiniApp可以通过<a href="https://html.spec.whatwg.org/multipage/webstorage.html#the-localstorage-attribute">window.localStorage</a>接口来访问这些数据。对于同一个<a href="https://www.w3.org/TR/miniapp-manifest/#dfn-app_id">app_id</a>的Widget和MiniApp，通过<a href="https://html.spec.whatwg.org/mul ...
```
```suggestion
      <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp Widget和MiniApp页面都可以通过user agent读写在本地保存的数据，没有过期时间。MiniApp Widget和MiniApp可以通过<a href="https://html.spec.whatwg.org/multipage/webstorage.html#the-localstorage-attribute">window.localStorage</a>接口来访问这些数据。对于同一个<a href="https://www.w3.org/TR/miniapp-manifest/#dfn-app_id">app_id</a>的Widget和MiniApp，通过<a href="https://html.spec.whatwg.org/multipage/webstorage.html#the-localstorage-attribute">window.localStorage</a>接口可以访问到相同的数据，示例如下：</p>
```

xfq commented on spec/index.html:
```diff
@@ -524,52 +514,102 @@ <h4>
-        <p its-locale-filter-list="en" lang="en">MiniApp Widget and MiniApp page can create and access user local files stored in the host environment, and can also access resource files in the application package. These two types of files can be located by several different types of URIs:</p>
+        <p its-locale-filter-list="zh-hans" lang="zh-hans">对每一个MiniApp，User agent 必须提供一个目录，用于存储user agent运行过程中产生的用户文件。MiniApp只可以在该目录下创建和访问文件，不能访问该目录以外的文件。MiniApp可以通过internal: URL访问私有文件。internal: URL 语法定义如下：
+</p>
+        <p its-locale-filter-list="en" lang="en">For each MiniApp, the User agent must provide a directory to store the user files generated during the operation of the user agent. MiniApp can only create and access files in that directory, not files outside of that directory. MiniApp can access private files via internal: URL.Internal: URL syntax is defined as follows:
```
```suggestion
        <p its-locale-filter-list="en" lang="en">For each MiniApp, the user agent must provide a directory to store the user files generated during the operation of the user agent. MiniApp can only create and access files in that directory, not files outside of that directory. MiniApp can access private files via internal: URL. Internal: URL syntax is defined as follows:
```

xfq commented on spec/index.html:
```diff
@@ -524,52 +514,102 @@ <h4>
-          internal://files/foo.txt
-          internal://tmp/foo.txt
-          internal://mass/foo.txt
+        <pre class="example js" title="">
```
```suggestion
        <pre class="example abnf" title="">
```

xfq commented on spec/index.html:
```diff
@@ -524,52 +514,102 @@ <h4>
+        </h4>
+		
+        <p its-locale-filter-list="zh-hans" lang="zh-hans">一些MiniApp Widget 的 UI components和API 可以通过属性和参数对文件进行访问，例如image 组件的的 src 属性：</p>
+        <p its-locale-filter-list="en" lang="en">MiniApp Widget and MiniApp page can access files through the src attribute of the tag and the File API，Example of tag src attribute:</p>
```
这里的中英文似乎不完全对应，中文提到的image组件和英文提到的File API都没有对应上。

lumanz replied:
谢谢！

xfq commented on spec/index.html:
```diff
@@ -524,52 +514,102 @@ <h4>
+        <p its-locale-filter-list="zh-hans" lang="zh-hans">一些MiniApp Widget 的 UI components和API 可以通过属性和参数对文件进行访问，例如image 组件的的 src 属性：</p>
+        <p its-locale-filter-list="en" lang="en">MiniApp Widget and MiniApp page can access files through the src attribute of the tag and the File API，Example of tag src attribute:</p>
+		
+		 <pre class="example js" title="Example: example of src attribute | src 属性示例">
```
```suggestion
		 <pre class="example xml" title="Example: example of src attribute | src 属性示例">
```

xfq commented on spec/index.html:
```diff
@@ -524,52 +514,102 @@ <h4>
+        <p its-locale-filter-list="en" lang="en">MiniApp Widget and MiniApp page can access files through the src attribute of the tag and the File API，Example of tag src attribute:</p>
+		
+		 <pre class="example js" title="Example: example of src attribute | src 属性示例">
+			<image src="platform://miniapp/foo/common/example.png" />
```
```suggestion
			&lt;image src="platform://miniapp/foo/common/example.png" /&gt;
```

这里需要转义。

xfq commented on spec/index.html:
```diff
@@ -524,52 +514,102 @@ <h4>
+        <span its-locale-filter-list="zh-hans" lang="zh-hans">通过文件API访问文件</span>
+        </h4>
 
+	        <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp可以通过 <a href="https://www.w3.org/TR/FileAPI/">File API</a>访问本地文件。</p>
+        <p its-locale-filter-list="en" lang="en">Access local files via the <a href="https://www.w3.org/TR/FileAPI/">File API</a>:</p>
```
```suggestion
	        <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp可以通过 [[FileAPI]] 访问本地文件。</p>
        <p its-locale-filter-list="en" lang="en">Access local files via [[FileAPI]]:</p>
```

xfq commented on spec/index.html:
```diff
@@ -578,33 +618,36 @@ <h4>
      <span its-locale-filter-list="en" lang="en">Security considerations</span>
       <span its-locale-filter-list="zh-hans" lang="zh-hans">安全考虑</span>  
       </h2>
+	    <p its-locale-filter-list="zh-hans" lang="zh-hans">由于widget的packaging、addressing和manifest与MiniApp page类似，因此widget的安全考虑宜参考MiniApp packaging、addressing和manifest中的安全考虑。此外：</p>
+    <p its-locale-filter-list="en" lang="en">Since the packaging, addressing, and manifest of widgets are similar to MiniApp pages, the security considerations of widgets should refer to the security considerations in MiniApp packaging, addressing, and manifest. Furthermore:</p>
```
```suggestion
	    <p its-locale-filter-list="zh-hans" lang="zh-hans">由于widget的packaging、addressing和manifest与MiniApp page类似，因此widget的安全考虑宜参考MiniApp <a href="https://www.w3.org/TR/miniapp-packaging/#sec-security">Packaging</a>、<a href="https://www.w3.org/TR/miniapp-addressing/#security-considerations">Addressing</a>和<a href="https://www.w3.org/TR/miniapp-manifest/#security-considerations">Manifest</a>中的安全考虑。此外：</p>
    <p its-locale-filter-list="en" lang="en">Since the packaging, addressing, and manifest of widgets are similar to MiniApp pages, the security considerations of widgets should refer to the security considerations in MiniApp <a href="https://www.w3.org/TR/miniapp-packaging/#sec-security">Packaging</a>, <a href="https://www.w3.org/TR/miniapp-addressing/#security-considerations">Addressing</a>, and <a href="https://www.w3.org/TR/miniapp-manifest/#security-considerations">Manifest</a>. Furthermore:</p>
```

lumanz committed:
Update index.html

lumanz committed:
Update index.html

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update spec/index.html

Co-authored-by: Fuqiao Xue <xfq@w3.org>

lumanz committed:
Update index.html

lumanz committed:
Merge branch 'main' of https://github.com/lumanz/miniapp-widget

lumanz committed:
Update introduction

add the relationship with other MiniApp specs

xfq commented on spec/index.html:
```diff
@@ -92,6 +92,13 @@ <h3>
+	  
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">本标准的内容符合MiniApp Manifest、Packaging、Lifecycle和Addressing的标准要求，除此之外，还规定了MiniApp特有的生命周期事件，以及MiniApp widget与MiniApp或与宿主环境间进行互动交流的要求。</p>
+  
+      <p its-locale-filter-list="en" lang="en">This specification follows the specifications of MiniApp Manifest, Packaging, Lifecycle and Addressing. In addition, it defines the unique life cycle events of MiniApp and the requirements for the interaction between MiniApp widget and MiniApp or with the host environment.</p>
```
```suggestion
      <p its-locale-filter-list="en" lang="en">This specification follows the specifications of MiniApp Manifest, Packaging, Lifecycle and Addressing. In addition, it defines the unique lifecycle events of MiniApp and the requirements for the interaction between MiniApp widget and MiniApp or with the host environment.</p>
```

xfq commented on spec/index.html:
```diff
@@ -237,10 +244,10 @@ <h2>
     <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp widget 沿用 MiniApp page的<a href="https://w3c.github.io/miniapp-lifecycle/#pagestate-enum">PageState enum</a>
       和<a href="https://w3c.github.io/miniapp-lifecycle/#miniapp-page-lifecycle-interface-and-callbacks"> 生命周期接口及回调</a>，
-      MiniApp widget 增加了 update 生命周期事件，具体描述在 6.4 MiniApp widget 更新过程处理。</p>
+      MiniApp widget 增加了 update 生命周期事件，具体描述在 5.4 MiniApp widget 更新过程处理。</p>
```
```suggestion
      MiniApp widget 增加了 update 生命周期事件，具体描述在[[[#update-process]]]处理。</p>
```

xfq commented on spec/index.html:
```diff
@@ -237,10 +244,10 @@ <h2>
     <p its-locale-filter-list="en" lang="en">MiniApp widget follows the <a href="https://w3c.github.io/miniapp-lifecycle/#pagestate-enum">PageState enum</a>
     and <a href="https://w3c.github.io/miniapp-lifecycle/#miniapp-page-lifecycle-interface-and-callbacks">lifecycle
-    interfaces and callbacks</a> of MiniApp page. The MiniApp widget adds the update lifecycle event, which is described in 6.4.</p>
+    interfaces and callbacks</a> of MiniApp page. The MiniApp widget adds the update lifecycle event, which is described in 5.4.</p>
```
```suggestion
    interfaces and callbacks</a> of MiniApp page. The MiniApp widget adds the update lifecycle event, which is described in [[[#update-process]]].</p>
```

xfq commented on spec/index.html:
```diff
@@ -594,7 +601,7 @@ <h4>
 		
         <p its-locale-filter-list="zh-hans" lang="zh-hans">一些MiniApp Widget 的 UI components和API 可以通过属性和参数对文件进行访问，例如image 组件的的 src 属性：</p>
-        <p its-locale-filter-list="en" lang="en">MiniApp Widget and MiniApp page can access files through the src attribute of the tag and the File API，Example of tag src attribute:</p>
+        <p its-locale-filter-list="en" lang="en">UI components and API of some MiniApp widget can access files through properties and parameters, such as the src property of the image component:</p>
```
```suggestion
        <p its-locale-filter-list="en" lang="en">UI components and API of some MiniApp widgets can access files through properties and parameters, such as the `src` attribute of the `image` component:</p>
```

xfq commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+<body>
+  <section id='abstract'>
+    <p its-locale-filter-list="en" lang="en">
+      The MiniApp Widget is a special form of <a href="https://www.w3.org/TR/miniapp-packaging/#dfn-page">MiniApp Page</a>.
```
```suggestion
      MiniApp Widget is a special form of <a href="https://www.w3.org/TR/miniapp-packaging/#dfn-page">MiniApp Page</a>.
```

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      <p its-locale-filter-list="en" lang="en">MiniApp widget is a special form of MiniApp page. Unlike pages, widgets can use only part of the screen (rather than all of
+      it) and are used to display key information to the user and respond to simple
+      user actions, such as displaying the weather at the user's location, tracking
+      user's current travel status and providing next actions .</p>
```
Suggestion: remove blank space before the period:
```
next actions .</p>
```

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">与 MiniApp page 类似，定位 MiniApp widget 及 Widget 中的资源的过程符合 <a
+    href="https://w3c.github.io/miniapp-addressing/">Miniapp Addressing</a> 定义的标准。</p>
+    <p its-locale-filter-list="en" lang="en">Similar to MiniApp page, the process of locating widget and the resources in it complies with the specification defined
+      by <a href="https://w3c.github.io/miniapp-addressing/">Miniapp Addressing</a>.</p>
```
Suggestion: https://www.w3.org/TR/miniapp-addressing/ instead of github.io.

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">与 page 不同的是， MiniApp widget 并不存在 start widget。因此，指定某一个 MiniApp widget 需要显式说明 Widget 的 path，如下例所示：</p>
+    <p its-locale-filter-list="en" lang="en">Unlike MiniApp pages, there is no start widget for a MiniApp widget. Therefore, specifying a MiniApp widget needs to
+    specify the path of the widget explicitly, as shown in Example:</p>
```
Suggestion: as shown in **the following example**:

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+    </h2>
+
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">包含 Widget 的 MiniApp 包结构符合如下要求：</p>
+    <p its-locale-filter-list="en" lang="en">The structure of MiniApp package containing a Widget should meets the following requirements:</p>
```
Suggestion: 
<p its-locale-filter-list="en" lang="en">The structure of **the** MiniApp package containing a Widget **SHOULD meet** the following requirements:</p>

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+
+    <p its-locale-filter-list="zh-hans" lang="zh-hans">与 MiniApp page 类似， MiniApp widget 相关的 metadata 也在 manifest.json 文件中定义，符合<a href="https://w3c.github.io/miniapp-manifest/">Miniapp Manifest</a>标准规范。</p>
+    <p its-locale-filter-list="en" lang="en">Similar to the MiniApp page, the metadata related to the MiniApp widget is also defined in the manifest.json file,
+      which meets to the <a href="https://w3c.github.io/miniapp-manifest/">Miniapp Manifest</a> specification.</p>
```
Suggestion:
```
meets the <a href="https://www.w3.org/TR/miniapp-manifest/">Miniapp Manifest</a> specification.</p>
```

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      <li>MiniApp widget loading</li>
+      <li>MiniApp widget update</li>
+      <li>MiniApp widget first render ready</li>
+      <li>MiniApp widget running in foreground</li>
```
Suggestion:
```
running in the foreground</li>
```

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      <li>MiniApp widget update</li>
+      <li>MiniApp widget first render ready</li>
+      <li>MiniApp widget running in foreground</li>
+      <li>MiniApp widget running in background</li>
```
Suggestion:
```
running in the background</li>
```

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+
+    <section id='loading-process'>
+      <h3>
+      <span its-locale-filter-list="en" lang="en"> Miniapp widget loading process</span>
```
Suggestion: `MiniApp`

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      </h3>
+
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">User agent 加载 MiniApp widget 并显示的过程如下：</p>
+      <p its-locale-filter-list="en" lang="en">The process of loading and displaying MiniApp widget by User agent is as follows: </p>
```
Suggestion: ` by a user agent is as follows: </p>`

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      </h3>
+
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">User agent 将 Widget 从界面移除并不再使用的情况下，会触发关闭过程:</p>
+      <p its-locale-filter-list="en" lang="en">When user agent removes the widget from the interface and is no longer using it, it will trigger the unloading process:</p>
```
Suggestion: `When **a** user agent`

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">MiniApp widget 相比于 MiniApp page ，会因为更多的原因不可见，这时 MiniApp widget 的
+        running in background 事件会被触发，例如：</p>
+      <p its-locale-filter-list="en" lang="en">Compared with MiniApp page, MiniApp widget will be invisible for more
+        reasons. The the running in background event of MiniApp widget will be triggered, for example: </p>
```
Please, remove duplicated `the`

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      </ol>
+        
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">当 MiniApp widget 重新可见时，running in foreground 事件会被触发。</p>    
+      <p its-locale-filter-list="en" lang="en">The running in foreground event would be triggered when MiniApp widget becomes visible again.</span></p>
```
Suggestion:
```
 <p its-locale-filter-list="en" lang="en">The running in foreground event would be triggered when **the** MiniApp widget becomes visible again.</span></p>
```

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">User agent 可以在不重新创建 Widget 环境的情况下更新 widget 的内容，这种情况常应用于 Widget 的 Host environment 与 Widget 有强交互的情况，
+        例如，在商品搜索结果页显示一个商品 widget 时，用户点击另一个商品，该 Widget 的内容将被更新为新商品的。</p>
+      <p its-locale-filter-list="en" lang="en">User agent can update the content of the
```
Suggestion: **A** user agent

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      <p its-locale-filter-list="zh-hans" lang="zh-hans">User agent 可以在不重新创建 Widget 环境的情况下更新 widget 的内容，这种情况常应用于 Widget 的 Host environment 与 Widget 有强交互的情况，
+        例如，在商品搜索结果页显示一个商品 widget 时，用户点击另一个商品，该 Widget 的内容将被更新为新商品的。</p>
+      <p its-locale-filter-list="en" lang="en">User agent can update the content of the
+        widget without recreating the widget environment. This often happens where the
```
Suggestion: This often happens **when** the

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+      <p its-locale-filter-list="en" lang="en">User agent can update the content of the
+        widget without recreating the widget environment. This often happens where the
+        widget host environment has strong interaction with the widget. For example,
+        when a product widget is displayed on search resualts page, user clicks on
```
Suggestion: the search **results** page

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+
+    <ul its-locale-filter-list="en" lang="en">
+      <li>Keeping the same login in state in widget when already logged in in MiniApp. </li>
+      <li>Widget gets playback status and display playback control when MiniApp is playing music. </li>
```
Suggestion: and display**s**

espinr commented on spec/index.html:
```diff
@@ -0,0 +1,691 @@
+	</ul>
+	
+	<p its-locale-filter-list="zh-hans" lang="zh-hans">针对不同类型的文件，可以采用多种URL形式来指定。</p>
+        <p its-locale-filter-list="en" lang="en">For different types of files, multiple URL forms cpuld be used.</p>
```
Suggestion: could (or MAY) be used

espinr reviewed (commented):
Please, find some editorial suggestions.
